Feature/pc 22 icons for tags

CuriousCactus opened the pull request:
Add more spell tags and corresponding svgs
Do tables within spells with proper markup
Do lists within spells with proper markup
Fix some spell typos

CuriousCactus commented:
The categories could probably be improved on (/rules-lawyered over), but this is a general idea of the approach.
It might be nice to give a description of what the categories mean somewhere, although I hope it's fairly obvious from the one-word label.
Some spells have multiple categories, which clutters the UI. We've discussed ways to fix this, but perhaps this should be a separate PR as there is a lot in this one already.
We could merge with multiple icons showing, or none. There would still be extra functionality from the filters (side-note - need to get the filters to display alphabetically).

prophet-kn commented on src/styles/_main.scss:
```diff
@@ -553,8 +564,8 @@
 
-  &-damage-icon {
-    margin-right: 15px;
+  &-damage-icon, &-category-icon {
```
New line here please

prophet-kn commented on src/data/spells.json:
```diff
@@ -399,11 +397,11 @@
     ],
-    "s_damage_dice": "None",
-    "s_damage_type": "None",
+    "s_damage_dice": "5",
```
5? isn't it 5 per level/spell level? should clarify

CuriousCactus replied:
I haven't included what happens at higher levels for any of the other spells, I think this is consistent.

prophet-kn replied:
Well, it looks lame with a flat 5. Also, doesn't it deal cold damage?

CuriousCactus replied:
It does deal cold damage. For me it shows:
**Damage:** 5 cold

prophet-kn commented on src/components/items/spell-item.js:
```diff
@@ -52,12 +52,22 @@ class SpellItem extends Component {
+
+    for (const [index, value] of spellItem.s_type.entries()) {
+      if (value !== "damage" && value !== "support" && value !== "utility") {
+        icons.push(<ReactSVG src={"svgs/"+value+".svg"} className="item-category-icon" beforeInjection={svg => {svg.setAttribute('style', 'height: 28px');}} key={index}/>)
```
Please add spaces between concatenating strings, makes it hard to read.

CuriousCactus replied:
Done.
Ideally we want linting so you're warned in the server terminal I guess?

prophet-kn replied:
More of like code clarity and some coding standards. As for coding standards, nothing is defined, so would be hard to pick a proper lint.

CuriousCactus replied:
I did a lot of configuring linting for my other project, I think I could sort something sensible. I'll put another feature request.

prophet-kn commented on src/components/items/spell-item.js:
```diff
@@ -52,12 +52,22 @@ class SpellItem extends Component {
+        icons.push(<ReactSVG src={"svgs/"+value+".svg"} className="item-category-icon" beforeInjection={svg => {svg.setAttribute('style', 'height: 28px');}} key={index}/>)
+      }
+      if (value === "damage" && spellItem.s_damage_type !== "None") {
+        icons.push(<ReactSVG src={"svgs/"+spellItem.s_damage_type+".svg"} className="item-damage-icon" beforeInjection={svg => {svg.setAttribute('style', 'height: 28px');}} key={index}/>)
```
Same here

prophet-kn commented on src/styles/_main.scss:
```diff
@@ -538,6 +539,16 @@
       }
     }
+
+    table {
```
table? where? :o

CuriousCactus replied:
There is a table in the teleport spell, for example (I added the markup tags to make it easier to style, it was just done with spaces and new lines)

prophet-kn replied:
okies

prophet-kn commented:
> The categories could probably be improved on (/rules-lawyered over), but this is a general idea of the approach.
> It might be nice to give a description of what the categories mean somewhere, although I hope it's fairly obvious from the one-word label.
> Some spells have multiple categories, which clutters the UI. We've discussed ways to fix this, but perhaps this should be a separate PR as there is a lot in this one already.
> We could merge with multiple icons showing, or none. There would still be extra functionality from the filters (side-note - need to get the filters to display alphabetically).

- To be honest I am more worried about the icons, that some people upon looking at them might not realise what they represent. I.e. acid vs poison or various damage types can be misinterpreted. I don't really care about rules-lawyers based on custom-made categories, opinions will always be opinions.
- Cluttering the UI is indeed bad, we want it to be as clear to view as possible.
- Not all the filters, I hope? Just the items inside of the filters, as I want to keep the structure from more important on the top, such as levels and classes.

CuriousCactus commented:
1) Yeah, I just picked the most obvious from one free icons site. Which do you personally find most confusing? 

For "Acid" we could use a simplified version of this (it's the symbol used on corrosive substances IRL):
![image](https://user-images.githubusercontent.com/17615620/87515773-041a3d80-c685-11ea-8110-0c1974c7cc28.png)
Probably we should just use the hand part to make it more compact.

For "Various" we could have some sort of merging swirl of some of the other symbols?

2) See https://github.com/prophet-kn/dnd_spells_book/issues/54

3) Correct, I meant items within filters

prophet-kn commented:
> 1. Yeah, I just picked the most obvious from one free icons site. Which do you personally find most confusing?
> 
> For "Acid" we could use a simplified version of this (it's the symbol used on corrosive substances IRL):
> ![image](https://user-images.githubusercontent.com/17615620/87515773-041a3d80-c685-11ea-8110-0c1974c7cc28.png)
> Probably we should just use the hand part to make it more compact.
> 
> For "Various" we could have some sort of merging swirl of some of the other symbols?
> 
> 1. See #54
> 2. Correct, I meant items within filters

Yeah, I don't mean specific examples, I am just saying that some people might confuse icons with other icons, and complain like 'but this doesn't represent acid damage erhmergherd' etc. :)

CuriousCactus commented:
All we can do is try to get the best representation I guess. I'll user test some on Oli.
We could also use words when the spell is expanded out, but again, clutter. For desktop we should use alt text, but that won't help on mobile.

CuriousCactus commented:
Things to do before merging:

Alt texts on icons (will only work on desktop)
User test icons, change as necessary

Anything else?

prophet-kn commented:
Another difference between desktop and mobile. :( Let's try to keep them as minimal as possible still. But yeah, sounds good, proceed. :) I haven't checked the code yet tho, only read the diff previously.

CuriousCactus commented:
User testing done (n=1 study though :P) and a few icons changed.

Tried to get tooltips on hovering over the icons, but got a very weird bug: https://github.com/tanem/react-svg/issues/871
I think this can be put in a separate PR if and when I work it out.

I've also turned the icons into a component, and moved cFC to a helper functions file so we don't have repeated code.

CuriousCactus commented:
Icon tooltips now fixed thanks to a speedy response from the maker of react-svg.

prophet-kn commented:
Ok, so, while I like the idea, we cannot have this due to the reasons shown below:

<img width="320" alt="Screenshot 2020-07-20 at 8 50 43" src="https://user-images.githubusercontent.com/45960358/87904159-664cb700-ca66-11ea-82d2-3d59967ba341.png">
<img width="320" alt="Screenshot 2020-07-20 at 8 50 55" src="https://user-images.githubusercontent.com/45960358/87904162-66e54d80-ca66-11ea-8d7f-9b938253a337.png">
<img width="320" alt="Screenshot 2020-07-20 at 8 51 09" src="https://user-images.githubusercontent.com/45960358/87904164-677de400-ca66-11ea-8516-e8bb33a88cff.png">
<img width="321" alt="Screenshot 2020-07-20 at 8 50 28" src="https://user-images.githubusercontent.com/45960358/87904165-677de400-ca66-11ea-996e-1fdf211b21af.png">
<img width="318" alt="Screenshot 2020-07-20 at 8 50 16" src="https://user-images.githubusercontent.com/45960358/87904167-677de400-ca66-11ea-899f-032e91503615.png">

User story:
Holy...4 icons, is it Holy Shit? is it Holy Inquisition? I dunno, I just wanted to look up my spell during combat, man, and now the DM is telling me my character is standing there dumbfounded and skips a turn just as I am looking for my Holy Aura spell.

On mobile these things tell me nothing, and even destroy the name and make it unreadable. Also, with so many icons loading, and I assume it loads them for each item as a new icon, clicking filters or burger menu is lagging, I am experiencing up to 1s delays, this doesn't happen without icons, please have a look at this again.

CuriousCactus commented:
What about only showing 1 icon? I can reorder the lists to put the most relevant one first.
Then when you open out the item you see all of them in a line somewhere in the main content.

I think/hope that the linting PR (https://github.com/prophet-kn/dnd_spells_book/pull/58) might help with performance. In this PR I'm loading the svgs from the 'public' folder, in the linting PR I think I'm doing it more 'properly', loading them with webpack from the src directory. Or it might just be that there are loads, in which case going down to rendering one per spell also makes sense. Some sort of lazy loading might also be possible, but let's try the other two ideas first.

prophet-kn commented:
That could be a solution, I'd say the most relevant icon would be the spell school type then?

Linting has been merged (hence the plethora of conflicts here), so you can try it. :)

CuriousCactus commented:
I don't have icons for spell school, they're separate categories.

prophet-kn commented:
Right, but if you think about it, we display a name and a level, from all of the possible other things that would be good to display as an icon (not limiting to the ones you've introduced), wouldn't school type be the most important?

CuriousCactus commented:
School doesn't matter for most classes. I think it's a hangover from old
editions. They're sort of useful classification, but not great. I'd rather
pick from the ones I've introduced.

On Mon, 20 Jul 2020, 10:09 prophet, <notifications@github.com> wrote:

> Right, but if you think about it, we display a name and a level, from all
> of the possible other things that would be good to display as an icon (not
> limiting to the ones you've introduced), wouldn't spell school type be the
> most important?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/prophet-kn/dnd_spells_book/pull/53#issuecomment-660846904>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEGMWBFP7J4DKQIRLHX3NBTR4PUUJANCNFSM4OZXR7OA>
> .
>

prophet-kn commented:
Maybe. Let's see which you deem worthy of being the main one then. :)

CuriousCactus commented:
Merge conflicts fixed.
Showing only one item on smaller screens implemented.

Loading icons is still a bit slow, and I think they aren't cached by the browser (probably because they're inline SVGs but I need to work this out because they _are_ being loaded from files, so in principal I would guess the files could be cached).

Also I think the Build Test is failing because it's running npm install when we're using yarn, but I need to look into that.

prophet-kn commented on src/components/icons/icons.js:
```diff
@@ -0,0 +1,25 @@
+import Icon from './icon'
+
+class Icons extends React.Component {
+  render() {
```
This is nice, but you know what would make it nicer? :) Having logic and loops be outside of render function, but rather be called in render from outside of it, so we'd have
```
render() {
  return (
    <span className={'item-icons'}>
      { this.getIcons() } // Where getIcons is the function outside of the render function.
    </span>
  )
}
```

CuriousCactus commented:
On checking, the icons it looks like the icons are cached, so it looks like the injection is a bit slow.

I've fixed the Build Test issue and moved the icons loop out of the render function.

prophet-kn commented on src/components/items/spell-item.js:
```diff
@@ -52,21 +52,11 @@ class SpellItem extends Component {
         <div className={'item-name'} onClick={(e) => { this.addClassName(e, i) }}>
           <span className={'item-label'}>{spellItem.s_name}</span>
-          {spellItem.s_damage_type !== 'None' && <ReactSVG src={svgs[spellItem.s_damage_type]} className="item-damage-icon" beforeInjection={svg => { svg.setAttribute('style', 'height: 28px') }}/>}
+          <Icons item={spellItem}/>
```
This is a nice component creation!

prophet-kn commented:
> On checking, the icons it looks like the icons are cached, so it looks like the injection is a bit slow.
> 
> I've fixed the Build Test issue and moved the icons loop out of the render function.

Yeah, there's still a small delay when entering the spell list page, and also when typing/filtering.
I might merge it already, unless you have an idea how to make it faster? I wonder if some sort of lazy-loading could be introduced or maybe if the reactsvg is actually what is causing the delay. Because it's the SVGs loading which is causing this delay, as you can see, on the 800ms:
<img width="602" alt="Screenshot 2020-08-03 at 11 47 16" src="https://user-images.githubusercontent.com/45960358/89163986-28c84d80-d57f-11ea-957a-cc68a46d8e37.png">

CuriousCactus commented:
I agree that it's in a good enough state that we could release it and improve later.
The main thing that isn't done is re-ordering the tags so that the most relevant icon is shown on mobile.

I guess we might see an improvement on small screens by only injecting the icon that will be shown (at the moment they're all injected, then all but the first are hidden with css).

Lazy loading is another possibility, but can we tell whether it's the injection or the rendering of the image that is slow? Could lazy injection be a thing??

prophet-kn commented:
I think it's the rendering, as i think (not sure) but it renders/loads for each item separately, which is why the delay does not exist when you have a slightly filtered list. Hm.
